packages/db/src/meta/pouch/workspace.ts

```diff
@@ -0,0 +1,227 @@
+    if (typeof id === "undefined") {
+      return;
+    }
+
+    const [savedRecord] = await this.adapter.retrieve<N, Input<C, N>>(
+      collectionName,
+      [{ _id: id } as RecordReference<Input<C, N>>]
+    );
+
+    if (savedRecord) {
+      return this.unmarshal(collectionName, savedRecord);
+    }
+  }
+
+  public async add<N extends CollectionName<C>>(
+    collectionName: N,
+    input: MutationInput<C, N>
+  ): Promise<MutationPayload<C, N>> {
+    const log = debug.extend(`${collectionName}:add`);
+    log("Adding...");
+
+    const records = input[collectionName].map(input =>
+      input ? this.marshal(collectionName, input) : undefined
+    );
+
+    const savedRecords = await this.adapter.record<N, Input<C, N>>(
+      collectionName,
+      records,
+      { overwrite: false }
+    );
+
+    const addedInputs = savedRecords.map(savedRecord =>
+      savedRecord ? this.unmarshal(collectionName, savedRecord) : undefined
+    );
+
+    log(
+      "Added ids: %o",
+      addedInputs
+        .filter((input): input is SavedInput<C, N> => !!input)
+        .map(({ id }) => id)
+    );
+
+    return {
+      [collectionName]: addedInputs
+    } as MutationPayload<C, N>;
+  }
+
+  public async update<M extends MutableCollectionName<C>>(
+    collectionName: M,
+    input: MutationInput<C, M>
+  ): Promise<MutationPayload<C, M>> {
+    const log = debug.extend(`${collectionName}:update`);
+    log("Updating...");
+
+    const records = input[collectionName].map(input =>
+      input ? this.marshal(collectionName, input) : undefined
+    );
+
+    const savedRecords = await this.adapter.record<M, Input<C, M>>(
+      collectionName,
+      records,
+      { overwrite: true }
+    );
+
+    const updatedInputs = savedRecords.map(savedRecord =>
+      savedRecord ? this.unmarshal(collectionName, savedRecord) : undefined
+    );
+
+    log(
+      "Updated ids: %o",
+      updatedInputs
+        .filter((input): input is SavedInput<C, M> => !!input)
+        .map(({ id }) => id)
+    );
+
+    return {
+      [collectionName]: updatedInputs
+    } as MutationPayload<C, M>;
+  }
+
+  public async remove<M extends MutableCollectionName<C>>(
+    collectionName: M,
+    input: MutationInput<C, M>
+  ): Promise<void> {
+    const log = debug.extend(`${collectionName}:remove`);
+    log("Removing...");
+
+    const records = input[collectionName].map(input =>
+      input ? this.marshal(collectionName, input) : undefined
+    );
+
+    await this.adapter.forget(collectionName, records);
+
+    log("Removed.");
+  }
+
+  private marshal<N extends CollectionName<C>>(
+    collectionName: N,
+    input: Input<C, N>
+  ): Record<Input<C, N>> {
+    const id = this.generateId<N>(collectionName, input);
+
+    if (typeof id === "undefined") {
+      throw new Error(`Unable to generate ID for "${collectionName}" resource`);
+    }
+
+    return {
+      ...input,
+      _id: id
+    };
+  }
+
+  private unmarshal<N extends CollectionName<C>>(
```

Comment:
This is a cool way to do this, I feel like!

Reply:
I'm proud of this pair! 😄